src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,140 @@
+        if (tail == null) {
+            head = newNode;
+        } else {
+            tail.setNext(newNode);
+        }
+        tail = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexForAdd(index);
+        if (index == size) {
+            add(value);
+        } else {
+            Node<T> current = getNode(index);
+            Node<T> newNode = new Node<>(current.getPrev(), value, current);
+            if (current.getPrev() == null) {
+                head = newNode;
+            } else {
+                current.getPrev().setNext(newNode);
+            }
+            current.setPrev(newNode);
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return getNode(index).getValue();
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = getNode(index);
+        T oldValue = current.getValue();
+        current.setValue(value);
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = getNode(index);
+        T value = current.getValue();
+        unlink(current);
+        return value;
     }
 
     @Override
     public boolean remove(T object) {
+        for (Node<T> current = head; current != null; current = current.getNext()) {
+            if (object == current.getValue() || (object != null
+                    && object.equals(current.getValue()))) {
+                unlink(current);
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> getNode(int index) {
+        Node<T> current;
+        if (index < (size >> 1)) {
+            current = head;
+            for (int i = 0; i < index; i++) {
+                current = current.getNext();
+            }
+        } else {
+            current = tail;
+            for (int i = size - 1; i > index; i--) {
+                current = current.getPrev();
+            }
+        }
+        return current;
+    }
+
+    private void unlink(Node<T> node) {
+        Node<T> next = node.getNext();
+        Node<T> prev = node.getPrev();
+
+        if (prev == null) {
+            head = next;
+        } else {
+            prev.setNext(next);
+            node.setPrev(null);
+        }
+
+        if (next == null) {
+            tail = prev;
+        } else {
+            next.setPrev(prev);
+            node.setNext(null);
+        }
+
+        node.setValue(null);
```

Comment:
Direct field access should be used instead of setters for the inner class `Node`. Since `Node` should be an inner class, you can access its fields directly if they are declared private.